### DISCUSSION TITLE
srsRAN 5G SA + Open 5GS + BladeRF 2 xA4

### DISCUSSION THREAD

**Discussion by MamdouhEgy** [srsran/srsRAN_Project#364] [General Help] [answered]
Hello All,

I am creating a new project to emulate a 5G SA network.
My hardware specs are:
 1- Dell OptiPlex Tower Plus 7010 (I7-13700K, 32GB RAM) --> to run the core and the installed srsRAN.
 2- BladeRF 2.0 micro xa4 --> as BS.
 3- Xiaomi 11 Lite 5G NE --> as UEs

My software specs:
1- Ubuntu 20.04.6, in which I installed the following [srsRAN](https://github.com/srsran/srsRAN_Project), [Open5GS](https://open5gs.org/open5gs/docs/guide/01-quickstart/) and [BladeRF2](https://github.com/Nuand/bladeRF). I already run the [srsran_performance](https://github.com/srsran/srsRAN_Project/tree/main/scripts/srsran_performance) script for best performance.

I followed this guide:
https://docs.srsran.com/projects/project/en/latest/tutorials/source/cotsUE/source/index.html

running **lsusb** gives the following:
        user@pc:~/srsRAN_Project/build/apps/gnb$ lsusb
        Bus 002 Device 002: ID 2cf0:5250 Nuand bladeRF 2.0

the output of **bladeRF-cli -e info -e version** command is attached as 
[bladeRFCommand.txt](https://github.com/srsran/srsRAN_Project/files/13602091/bladeRFCommand.txt)

Th output of **ip address show dev lo** command is attached as  [lo-interface.txt](https://github.com/srsran/srsRAN_Project/files/13602086/lo-interface.txt)

The configuration of the gnb_rf.yml which located in /home/user/srsRAN_Project/build/apps/gnb is attached as
[gnb_rf.txt](https://github.com/srsran/srsRAN_Project/files/13602128/gnb_rf.txt).

The configuration of the amf.yaml which located in /etc/open5gs is attached as 
[amf.txt](https://github.com/srsran/srsRAN_Project/files/13602148/amf.txt)

The configuration of the upf.yaml which located in /etc/open5gs is attached as 
[upf.txt](https://github.com/srsran/srsRAN_Project/files/13602154/upf.txt)

I get this error always which i think something wrong with the installation or drivers... any clue?
![Screenshot from 2023-12-07 16-52-23](https://github.com/srsran/srsRAN_Project/assets/42903127/18760f96-c1f9-4d2e-80d5-e4e02054caa5)

the proposed solutions here in this [Discussion](https://github.com/srsran/srsRAN_Project/issues/181) didnot help.

I also have a confusion related to the IPs (_in my PC_, in _amf.yaml_ , in _gnb_rf.yml_ and in _upf.yaml_)also

_amf.yaml:_
ngap:
    server:
        address: 127.0.0.1

_gnb_rf.yml:_
    amf:
        addr: 127.0.0.1
        bind_addr: 127.0.0.1
        
        
_upf.yaml_
     gtpu:
        server:
            - address: 127.0.1.100
        
        
I know this seems alot to read but iam totally new to such topic and try to figure out what is going on.

Best Regards
Mamdouh

**Comment by brendan-mcauliffe** [comment, marked as answer]
Hi @MamdouhEgy the issue is that you are trying to use a BladeRF SDR. Currently, only UHD is supported as the RF driver. If you can, you will need to use a USRP as the RF frontend.

**Reply by MamdouhEgy** [reply]
Thanks @brendan-mcauliffe.
So according to [5g_sa_COTS](https://docs.srsran.com/projects/4g/en/latest/app_notes/source/5g_sa_COTS/source/index.html) and [SRS RAN 5G SA with bladerf 2.0](https://github.com/srsran/srsRAN_4G/issues/896), my current setup will work if I just replaced my srsRAN 5G SA with srsRAN 4G 22.04 (or later)? I want to avoid buying USRP out of cost.

**Reply by brendan-mcauliffe** [reply]
In theory yes, although we do not recommend using srsRAN 4G for rigorous 5G testing and implementation. The 5G features in srsRAN 4G are intended for prototyping and PoCs. You could also use the Soapy -> UHD driver, this is not something we have tested, but other users have used this. You can find a post about it on the discussion board.

**Comment by maxbraunschmied** [comment]
Did you find a solution/get it to work? I am trying to establish a 5G SA Network with the same bladeRF.

**Reply by brendan-mcauliffe** [reply]
Hi @maxbraunschmied, as I have previously said, srsRAN Project **does not support the Soapy driver** required by the BladeRF. You can try to use the Soapy to UHD driver (SoapyUHD). Although we have not tested this in-house so cannot guarantee performance or support. 

You can read about SoapyUHD here: https://github.com/pothosware/SoapyUHD

**Comment by JPwnage** [comment]
AliExpress has some B200 (usrp clones) for around the same price as a blade RF  2.0 Xa4. 
Between $550-$600. 

My suggestion is return or Sell the blade and try one of those. (Although an authentic USRP would be best. But like you, I couldn't afford one either, my clone works great though.)

**Reply by orion-belt** [reply]
hi @JPwnage 
Can you provide the Aliexpress link for B200 clone you purchased & which worked for you ?

**Comment by JPwnage** [comment]
Another note, if you can't develop yourself, you may be able to find a developer on a freelance site (fiverr) who can to help port the driver to make it compatible with the bladeRF. 
( Take my advice here with a grain of salt)

**Comment by brendan-mcauliffe** [comment]
You can take a look at [this post](https://github.com/srsran/srsRAN_Project/discussions/222#discussioncomment-7120314), some other users have clearly shown how to get srsRAN Project to interface with bladeRF.

**Reply by maxbraunschmied** [reply]
Thanks, I got it to work via this thread! :)
